File: electron_node/electron-node/main/src/phonetic-correction/rescore.ts
/**
 * 同音候选 + LM 打分 + delta 选优。单一路径：候选生成 → 打分 → 阈值选优 → 返回文本。
 */

import { getLmScorer } from './lm-scorer';
import { getReplaceablePositions, generateCandidates } from './confusion-set';

export interface RescoreOpts {
  maxPositions: number;
  maxCandidates: number;
  delta: number;
}

const DEFAULT_OPTS: RescoreOpts = {
  maxPositions: 2,
  maxCandidates: 24,
  delta: 1.0,
};

export interface RescoreResult {
  text: string;
  changed: boolean;
  deltaScore?: number;
  candidatesCount?: number;
}

const MIN_LEN = 2;
const MAX_LEN = 120;

/**
 * 候选生成 → LM 打分 → 仅当 score(best) - score(original) >= delta 时替换。
 * LM 不可用或无可替换位点时返回原文（fail-open）。
 */
export async function rescoreWithLm(
  text: string,
  opts: Partial<RescoreOpts> = {}
): Promise<RescoreResult> {
  const t = text.trim();
  if (!t || t.length < MIN_LEN || t.length > MAX_LEN) {
    return { text, changed: false };
  }

  const scorer = getLmScorer();
  if (!scorer) return { text, changed: false };

  const { maxPositions, maxCandidates, delta } = { ...DEFAULT_OPTS, ...opts };
  const positions = getReplaceablePositions(t, maxPositions);
  if (positions.length === 0) return { text, changed: false };

  const candidates = generateCandidates(t, positions, maxCandidates);
  const origScore = (await scorer.score(t)).score;
  let bestText = t;
  let bestScore = origScore;

  for (let i = 1; i < candidates.length; i++) {
    const s = (await scorer.score(candidates[i])).score;
    if (s > bestScore) {
      bestScore = s;
      bestText = candidates[i];
    }
  }

  const deltaScore = bestScore - origScore;
  if (deltaScore < delta) return { text, changed: false, deltaScore, candidatesCount: candidates.length };
  return { text: bestText, changed: true, deltaScore, candidatesCount: candidates.length };
}

File: electron_node/services/phonetic_correction_zh/core/rescore.py
# -*- coding: utf-8 -*-
"""同音候选 + LM 打分 + delta 选优。"""

from typing import Optional

from .confusion_set import get_replaceable_positions, generate_candidates
from .lm_scorer import get_lm_scorer

MIN_LEN = 2
MAX_LEN = 120
DEFAULT_MAX_POSITIONS = 2
DEFAULT_MAX_CANDIDATES = 24
DEFAULT_DELTA = 1.0


def rescore_with_lm(
    text: str,
    service_dir: str,
    max_positions: int = DEFAULT_MAX_POSITIONS,
    max_candidates: int = DEFAULT_MAX_CANDIDATES,
    delta: float = DEFAULT_DELTA,
) -> tuple:
    """候选生成 → LM 打分 → 仅当 score(best) - score(original) >= delta 时替换。"""
    t = text.strip()
    if not t or len(t) < MIN_LEN or len(t) > MAX_LEN:
        return text, False, None, None
    scorer = get_lm_scorer(service_dir)
    if scorer is None:
        return text, False, None, None
    positions = get_replaceable_positions(t, max_positions)
    if not positions:
        return text, False, None, None
    candidates = generate_candidates(t, positions, max_candidates)
    orig_score, _ = scorer.score(t)
    best_text = t
    best_score = orig_score
    for i in range(1, len(candidates)):
        s, _ = scorer.score(candidates[i])
        if s > best_score:
            best_score = s
            best_text = candidates[i]
    delta_score = best_score - orig_score
    if delta_score < delta:
        return text, False, delta_score, len(candidates)
    return best_text, True, delta_score, len(candidates)


def phonetic_correct(
    text: str,
    service_dir: str,
    max_positions: int = DEFAULT_MAX_POSITIONS,
    max_candidates: int = DEFAULT_MAX_CANDIDATES,
    delta: float = DEFAULT_DELTA,
) -> str:
    """唯一入口：有 LM 则 rescore 选优，无 LM 或无可替换位点则返回原文。"""
    t = text.strip()
    if not t:
        return text
    result_text, _, _, _ = rescore_with_lm(
        text, service_dir, max_positions, max_candidates, delta
    )
    return result_text
